Intel UGW Software addon:-
Add ip-pre-down script support. Also added support to pass it over commandline.

--- a/pppd/ipcp.c
+++ b/pppd/ipcp.c
@@ -1992,6 +1992,9 @@ ipcp_down(f)
     if (demand) {
 	sifnpmode(f->unit, PPP_IP, NPMODE_QUEUE);
     } else {
+	/* run the pre-down script, if any, and wait for it to finish */
+	ipcp_script(path_ippredown, 1);
+
 	sifnpmode(f->unit, PPP_IP, NPMODE_DROP);
 	sifdown(f->unit);
 	ipcp_clear_addrs(f->unit, ipcp_gotoptions[f->unit].ouraddr,
--- a/pppd/pathnames.h
+++ b/pppd/pathnames.h
@@ -25,6 +25,7 @@
 #define _PATH_IPUP	 _ROOT_PATH "/etc/ppp/ip-up"
 #define _PATH_IPDOWN	 _ROOT_PATH "/etc/ppp/ip-down"
 #define _PATH_IPPREUP	 _ROOT_PATH "/etc/ppp/ip-pre-up"
+#define _PATH_IPPREDOWN	 _ROOT_PATH "/etc/ppp/ip-pre-down"
 #define _PATH_AUTHUP	 _ROOT_PATH "/etc/ppp/auth-up"
 #define _PATH_AUTHDOWN	 _ROOT_PATH "/etc/ppp/auth-down"
 #define _PATH_TTYOPT	 _ROOT_PATH "/etc/ppp/options."
--- a/pppd/main.c
+++ b/pppd/main.c
@@ -320,6 +320,7 @@ main(argc, argv)
 
     strlcpy(path_ipup, _PATH_IPUP, sizeof(path_ipup));
     strlcpy(path_ipdown, _PATH_IPDOWN, sizeof(path_ipdown));
+    strlcpy(path_ippredown, _PATH_IPPREDOWN, sizeof(path_ippredown));
     strlcpy(path_ipv6up, _PATH_IPV6UP, sizeof(path_ipv6up));
     strlcpy(path_ipv6down, _PATH_IPV6DOWN, sizeof(path_ipv6down));
 
--- a/pppd/options.c
+++ b/pppd/options.c
@@ -117,6 +117,7 @@ int	connect_delay = 1000;	/* wait this m
 int	req_unit = -1;		/* requested interface unit */
 char	path_ipup[MAXPATHLEN];	/* pathname of ip-up script */
 char	path_ipdown[MAXPATHLEN];/* pathname of ip-down script */
+char	path_ippredown[MAXPATHLEN]; /* pathname of ip-pre-down script */
 char	path_ipv6up[MAXPATHLEN];	/* pathname of ipv6-up script */
 char	path_ipv6down[MAXPATHLEN];/* pathname of ipv6-down script */
 bool	multilink = 0;		/* Enable multilink operation */
@@ -303,6 +304,9 @@ option_t general_options[] = {
     { "ip-down-script", o_string, path_ipdown,
       "Set pathname of ip-down script",
       OPT_PRIV|OPT_STATIC, NULL, MAXPATHLEN },
+    { "ip-pre-down-script", o_string, path_ippredown,
+      "Set pathname of ip-pre-down script",
+      OPT_PRIV|OPT_STATIC, NULL, MAXPATHLEN },
 
     { "ipv6-up-script", o_string, path_ipv6up,
       "Set pathname of ipv6-up script",
--- a/pppd/pppd.h
+++ b/pppd/pppd.h
@@ -319,6 +319,7 @@ extern int	max_data_rate;	/* max bytes/s
 extern int	req_unit;	/* interface unit number to use */
 extern char	path_ipup[MAXPATHLEN]; /* pathname of ip-up script */
 extern char	path_ipdown[MAXPATHLEN]; /* pathname of ip-down script */
+extern char	path_ippredown[MAXPATHLEN]; /* pathname of ip-predown script */
 extern char	path_ipv6up[MAXPATHLEN]; /* pathname of ipv6-up script */
 extern char	path_ipv6down[MAXPATHLEN]; /* pathname of ipv6-down script */
 extern bool	multilink;	/* enable multilink operation */
